### PR TITLE
macOS: pump event queue when waiting on fences.

### PR DESCRIPTION
**Pull request by prideout**

### CODE REVIEW

File: filament/backend/src/opengl/PlatformCocoaGL.mm [hunk 1]
```diff
@@ -144,6 +144,21 @@
     [pImpl->mGLContext flushBuffer];
 }
 
+bool PlatformCocoaGL::pumpEvents() noexcept {
+    if (![NSThread isMainThread]) {
+        return false;
+    }
+    while (true) {
+        NSEvent *event = [NSApp nextEventMatchingMask:NSEventMaskAny untilDate:[NSDate distantPast] inMode:NSDefaultRunLoopMode dequeue:YES];
```
Comment (bejado) [inline filament/backend/src/opengl/PlatformCocoaGL.mm]: Seems like the `untilDate` argument should be `[NSDate distantFuture]`?

**Comment by prideout** [comment]
@pixelflinger  this is ready for your review, I did a test and it seems to actually fix the deadlock!

**Comment by prideout** [comment]
I replaced the sleep with a non-zero timeout and it still works.

File: filament/src/Fence.cpp [hunk 2]
```diff
@@ -77,7 +82,29 @@ FenceStatus FFence::wait(Mode mode, uint64_t timeout) noexcept {
...
+        const auto startTime = std::chrono::system_clock::now();
+        while (true) {
+            status = fs->wait(ns(ms(PUMP_INTERVAL_MILLISECONDS)).count());
+            if (status == FenceStatus::CONDITION_SATISFIED) {
+                break;
+            }
+            mEngine.pumpPlatformEvents();
+            const auto elapsed = std::chrono::system_clock::now() - startTime;
+            if (elapsed >= ns(timeout)) {
+                break;
```
Comment (pixelflinger) [inline filament/src/Fence.cpp]: shouldn't you set `status` to `TIMEOUT_EXPIRED` here?

File: filament/src/Fence.cpp [hunk 3]
```diff
@@ -77,7 +82,29 @@ FenceStatus FFence::wait(Mode mode, uint64_t timeout) noexcept {
...
+        // Unfortunately, some platforms might force us to have sync points between the GL thread
+        // and user thread. To prevent deadlock on these platforms, we chop up the waiting time into
+        // polling and pumping the platform's event queue.
+        const auto startTime = std::chrono::system_clock::now();
+        while (true) {
+            status = fs->wait(ns(ms(PUMP_INTERVAL_MILLISECONDS)).count());
+            if (status == FenceStatus::CONDITION_SATISFIED) {
+                break;
+            }
+            mEngine.pumpPlatformEvents();
```
Comment (pixelflinger) [inline filament/src/Fence.cpp]: store `mEngine` to a local, so we don't have to dereference `this` each time.

File: filament/src/Fence.cpp [hunk 4]
```diff
@@ -77,7 +82,29 @@ FenceStatus FFence::wait(Mode mode, uint64_t timeout) noexcept {
...
+    if (UTILS_LIKELY(!mEngine.pumpPlatformEvents())) {
+        status = fs->wait(timeout);
+    } else {
+        // Unfortunately, some platforms might force us to have sync points between the GL thread
+        // and user thread. To prevent deadlock on these platforms, we chop up the waiting time into
+        // polling and pumping the platform's event queue.
+        const auto startTime = std::chrono::system_clock::now();
+        while (true) {
+            status = fs->wait(ns(ms(PUMP_INTERVAL_MILLISECONDS)).count());
+            if (status == FenceStatus::CONDITION_SATISFIED) {
```
Comment (pixelflinger) [inline filament/src/Fence.cpp]:
```suggestion
            if (status != FenceStatus::TIMEOUT_EXPIRED) {
```

**Pull request merged**